### PR TITLE
Add the `NodeReference.neighbors` implementation

### PR DESCRIPTION
**Pull request by decentralion**
The implementation is adapted from our previous implementation, but has
been refactored to be more appropriate for our generator-function
approach.

The unit tests comprehensively explore a simple example graph, testing
the following cases:
- Direction filtering (`IN`/`OUT`/`ANY`/unspecified)
- Node types
- Edge types
- Self-loop edges
- Dangling edges
- Removed edges don't appear

Test plan: Carefully inspect the added unit tests.

Paired with @wchargin

### CODE REVIEW

File: src/core2/graph.test.js [hunk 1]
```diff
@@ -71,6 +71,128 @@ describe("graph", () => {
...
+      //    Foo  > (Foo)
+      //     V
+      //   (B404)
+      const b1 = new BarPayload(1, "hello");
+      const b404 = new BarPayload(404, "hello");
+      const foo = new FooPayload();
+      const b1_foo = edge("b1_foo", b1, foo);
+      const foo_foo = edge("foo_foo", foo, foo, "SELF");
+      const foo_b404 = edge("foo_b404", foo, b404);
+      const phantomNode = new BarPayload(666, "ghost");
```
Comment (wchargin) [inline src/core2/graph.test.js]: The terms that we've been using so far are:
  - *present*: node is in the graph
  - *phantom*: node has been added to and removed from the graph (and is not in the graph)
  - *absent*: node was never in the graph
  - *isolated* (not trichotomous): node has no incident edges

So, your `phantomNode` isn't actually a phantom node. Could you rename it or update the graph structure accordingly?

Comment (decentralion) [reply src/core2/graph.test.js]: Done

File: src/core2/graph.js [hunk 2]
```diff
@@ -160,6 +161,10 @@ export class Graph {
     if (!indexedEdge) {
       return undefined;
     }
+    return this._upgradeIndexedEdge(indexedEdge);
+  }
+
+  _upgradeIndexedEdge(indexedEdge: IndexedEdge): Edge<any> {
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core2/graph.js]: Done.

File: src/core2/graph.js [hunk 3]
```diff
@@ -433,10 +438,51 @@ class InternalReference implements NodeReference {
...
   ): Iterator<{|+ref: NodeReference, +edge: Edge<any>|}> {
-    const _ = options;
-    throw new Error("Not implemented");
+    const indexDatum = this._graph._nodeIndices.get(this._address);
+    if (indexDatum == null) {
+      return;
+    }
+    const nodeIndex = indexDatum.index;
+
+    const direction = (options && options.direction) || "ANY";
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core2/graph.js]: Done

File: src/core2/graph.test.js [hunk 4]
```diff
@@ -71,6 +71,138 @@ describe("graph", () => {
...
+      // It should be fine so long as neighbors never mutates the graph.
+      const edge = (id, src, dst, type = "EDGE") => ({
+        address: {id, owner: {plugin: EXAMPLE_PLUGIN_NAME, type}},
+        src: src.address(),
+        dst: dst.address(),
+        payload: {},
+      });
+      // A cute little diagram:
+      // A>B = edge from A to B
+      // Nodes in the graph: B1,Foo
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core2/graph.test.js]: Fixed

File: src/core2/graph.test.js [hunk 5]
```diff
@@ -71,6 +71,138 @@ describe("graph", () => {
...
+          .addNode(bar);
+        const ref = g.ref(foo.address());
+        expectNeighborsEqual(ref.neighbors(), []);
+      });
+      it("finds neighbors for an absent node", () => {
+        expectNeighborsEqual(refFor(absent).neighbors(), [
+          {edge: foo_absent, ref: refFor(foo)},
+        ]);
+      });
+      describe("it filters by direction:", () => {
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core2/graph.test.js]: Fixed

File: src/core2/graph.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core2/graph.js]: Update: Doing this gave me a typing headache, around properly expressing the set `{||} | {|+plugin|} | {|+plugin, +type|}`. Once I did that, it wasn't happy to assign the `{}` as a PluginFilter in the default neighbor options above.
I'm going to merge it as is, since the code is fine as written. Feel free to PR an improvement!

[the remark this answers is not in the document]
Comment (wchargin) [reply src/core2/graph.js]: Yep; that known issue is why I didn’t suggest it outright. I spent a bit
of time poking around in Flow’s inference engine to see if there might
be an easy way to infer `{}` as having either type `{||}` or the
unsealed object type, but didn’t find much useful. It wouldn’t be as
simple as inferring the intersection type `{} & {||}`; this would be
unsound. One interesting thing that I learned is that `flow type-at-pos`
in fact infers `{||}` as the type of `const emptyObject = {}`, which is
surprising given that `const emptyObject: {||} = {}` yields a type
error. But that’s not particularly actionable, as far as I can tell.

So, TL;DR: merging as-is is fine with me.
